src/System.Private.CoreLib/src/System/Threading/WaitHandle.Windows.cs

```diff
@@ -0,0 +1,298 @@
+        }
+
+        private static void ThrowAbandonedMutexException(int location, WaitHandle handle)
+        {
+            throw new AbandonedMutexException(location, handle);
+        }
+
+        internal static void ThrowSignalOrUnsignalException()
+        {
+            int errorCode = Interop.mincore.GetLastError();
+            switch (errorCode)
+            {
+                case Interop.mincore.Errors.ERROR_INVALID_HANDLE:
+                    throw InvalidOperationException.NewInvalidHandle();
+
+                case Interop.mincore.Errors.ERROR_TOO_MANY_POSTS:
+                    throw new SemaphoreFullException();
+
+                case Interop.mincore.Errors.ERROR_NOT_OWNER:
+                    throw new SynchronizationLockException();
+                    // TODO: netstandard2.0 - After switching to ns2.0 contracts, use the below instead for compatibility
+                    //throw new ApplicationException(SR.Arg_SynchronizationLockException);
+
+                default:
+                    var ex = new Exception();
+                    ex.SetErrorCode(errorCode);
+                    throw ex;
+            }
+        }
+
+        private static void ThrowWaitFailedException(int errorCode)
+        {
+            switch (errorCode)
+            {
+                case Interop.mincore.Errors.ERROR_INVALID_HANDLE:
+                    throw InvalidOperationException.NewInvalidHandle();
+
+                case Interop.mincore.Errors.ERROR_INVALID_PARAMETER:
+                    throw new ArgumentException();
+
+                case Interop.mincore.Errors.ERROR_ACCESS_DENIED:
+                    throw new UnauthorizedAccessException();
+
+                case Interop.mincore.Errors.ERROR_NOT_ENOUGH_MEMORY:
+                    throw new OutOfMemoryException();
+
+                case Interop.mincore.Errors.ERROR_TOO_MANY_POSTS:
+                    /// Only applicable to <see cref="WaitHandle.SignalAndWait(WaitHandle, WaitHandle)"/>. Note however, that
+                    /// if the semahpore already has the maximum signal count, the Windows SignalObjectAndWait function does not
+                    /// return an error, but this code is kept for historical reasons and to convey the intent, since ideally,
+                    /// that should be an error.
+                    throw new InvalidOperationException(SR.Threading_SemaphoreFullException);
+
+                case Interop.mincore.Errors.ERROR_NOT_OWNER:
+                    /// Only applicable to <see cref="WaitHandle.SignalAndWait(WaitHandle, WaitHandle)"/> when signaling a mutex
+                    /// that is locked by a different thread. Note that if the mutex is already unlocked, the Windows
+                    /// SignalObjectAndWait function does not return an error.
+                    throw new SynchronizationLockException();
+                    // TODO: netstandard2.0 - After switching to ns2.0 contracts, use the below instead for compatibility
```

Comment:
Switch now here and put the workaround in the test because that's where the actual problem is?